### ISSUE TITLE
CmiFree reference count was zero abort on XC30

### ISSUE THREAD

**Issue by trquinn** [UIUC-PPL/charm#546]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/546*

---
A ChaNGa run fails on "Piz Daint" with the following error:

```
Big step 1171 took 64.667949 seconds.
ReOrder particles ... took 2.669491 seconds.
Writing Tipsy file ... took 59.006577 seconds.
Writing arrays ... took 0.000007 seconds.
Starting big step 1172
Rung distribution: ( 36267769, 12448115, 10134489, 12951858, 4067274, 1334, 0, 0
, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,)
MaxRung: 5
Maximum memory: 2896 MB
Kick Open:
 Rung 0: 4.28953e-05
 Rung 1: 2.14477e-05
 Rung 2: 1.07238e-05
 Rung 3: 5.36192e-06
 Rung 4: 2.68096e-06
 Rung 5: 1.34048e-06
Drift: Rung 5 Delta 2.68096e-06

Step: 1171.031250 Time: 0.100970 Rungs 5 to 5. Gravity Active: 1334, Gas Active:
 0
Domain decomposition ... ------------- Processor 2044 Exiting: Called CmiAbort -
-----------
Reason: CmiFree reference count was zero-- is this a duplicate free?
[2044] Stack Traceback:
  [2044:0] [0x20295535]
  [2044:1] [0x2029fee2]
  [2044:2] [0x202965d6]
  [2044:3] [0x2029d113]
  [2044:4] [0x2029d29e]
```

The traceback points to:
0x202965d6 <PumpRemoteTransactions+182>
as the place where CmiFree() is called.

Steps to reproduce:

build charm version v6.6.0-rc4-0-g63de3ca with:
./build ChaNGa gni-crayxc hugepages smp -j4 -O2

build ChaNGa version v3.0-94-g3e38ac3 with:
./configure --enable-bigkeys; make 

Submit the attached ".job" script (fails on both 128 and 256 cores) using the attached ".param" file, and the Eris_Core.001170 dataset (currently being transferred to BlueWaters in ~trq/scratch/eris; also available on Piz Daint at: /users/trq/scratch_daint).  The crash will happen in a few minutes.  From the output, it seems that the crash is happening during a phase involving a lot of particle movement as the code rearranges the particles from "output order" to "tree order".

[simulare_ch.job](https://github.com/UIUC-PPL/charm/files/3536483/simulare_ch.job.txt)
[T_S.param](https://github.com/UIUC-PPL/charm/files/3536484/T_S.param.txt)

**Comment by PhilMiller** [comment]
*Original date: 2014-08-14 20:26:37*

---
For the sake of simplification, does this happen with charm built with `./build ChaNGa gni-crayxc hugepages -j4`

**Comment by trquinn** [comment]
*Original date: 2014-08-14 21:23:36*

---
With charm built with "./build ChaNGa gni-crayxc hugepages -j4 -O2" and changa
run with

aprun -n 128 -N 8 ./ChaNGa.nsmp -oi 1 -D 1 -p 65536 -wall 30 -killat 2000 T_S.param +balancer MultistepLB_notopo +pemap 0-7 +setcpuaffinity +LBCommOff +LBDebug 1 +noAnytimeMigration

It gets past that point and dies slightly further on with:
libhugetlbfs [nid04228:11639]: WARNING: New heap segment map at 0x100ac000000 failed: Cannot allocate memory

The memory use seems to indicate that this is expected (memory use in ChaNGa can be much higher without SMP).

I reran with 256 cores and it ran OK.

Bottom line (at this point) non-SMP seems to be OK.

**Comment by pplimport** [comment]
*Original author: Yanhua Sun*
*Original date: 2014-08-15 09:22:44*

---
Is this problem specific on 'Piz Daint' or also happening on other machines? I still have not got account on Piz Daint.  that is one reason that I am asking.

**Comment by trquinn** [comment]
*Original date: 2014-08-15 17:02:57*

---
It does not happen on Blue Waters.  I do not have an account on Edison, so I can't try it there.

**Comment by harshithamenon** [comment]
*Original date: 2014-08-18 22:25:29*

---
I tried it on Edison and the crash was not reproducible. I ran it on 192 cores (8 nodes) in SMP mode, gnu compiler.

**Comment by pplimport** [comment]
*Original author: Gengbin Zheng*
*Original date: 2014-09-03 20:23:49*

---
Tom, 
can you try with REMOTE_EVENT turned off?

**Comment by trquinn** [comment]
*Original date: 2014-09-05 19:18:25*

---
With
#define REMOTE_EVENT 0
in machine.c, the run does not crash.
As far as I can tell, there isn't a performance difference.

**Comment by jcphill** [comment]
*Original date: 2014-11-05 20:00:17*

---
When I try disabling REMOTE_EVENT and enabling persistent messages I get the following error:

machine.c(3276): error #165: too few arguments in function call
          status = send_smsg_message(queue, ptr->destNode, ptr->msg, ptr->size, ptr->tag, 1, ptr);  
                                                                                                ^

It appears that several calls to send_smsg_message() were not updated with the new "ischarm" final argument.

**Comment by pplimport** [comment]
*Original author: Yanhua Sun*
*Original date: 2014-11-05 20:12:42*

---
Sorry that I forgot to mention that several minor changes are required to make it compile. I will submit a patch for Charm to let persistent work without remote event

**Comment by pplimport** [comment]
*Original author: Yanhua Sun*
*Original date: 2014-11-07 18:16:55*

---
checked in  http://charm.cs.uiuc.edu/gerrit/425

**Comment by nikhil-jain** [comment]
*Original date: 2014-11-11 18:58:15*

---
Does there exist another issue which keep an open bug report, so that we can actually fix this bug in future, instead of disabling the remote events?
